### PR TITLE
HF-307 PR 2: public API guards (ensureCapability)

### PR DESCRIPTION
**Pull request by marcin-kordas-hoc**
> **Update 2026-09-22 — review round with @sequba.** The sections below are the PR as written
> before that round, kept for the reasoning they record. Where a sentence no longer matches the
> code it is corrected inline and marked, rather than rewritten away.

### Context

HF-307 (feature packages / entitlement gating). This is PR 2 of 4: guards on the public API. Stacked on hf-307-entitlement-gating-pr1 (#1728) - depends on `CapabilityRegistry`/`allowsFeature`/`FeatureId`/`Config.licenseCapabilities` from that PR. Mechanical, public-repo-only, no key-format knowledge needed.

New:
- `src/errors.ts` - `LicenseCapabilityMissingError`, mirroring the existing ~30 error classes. Now also exported from the public entrypoint (`src/index.ts`) - see finding 6.
- `src/HyperFormula.ts` - private `ensureCapability(feature: FeatureId)`, mirroring `ensureEvaluationIsNotSuspended`: a single `isLicenseGateActive` boolean read on the fast path, then `allowsFeature` against the resolved entitlement.
- `src/BuildEngineFactory.ts` - private static `ensureNamedExpressionsCapability(config, namedExpressions)`, the build-time counterpart (task 2.3).
- ~~`src/CrudOperations.ts` - public `isCutClipboard()`~~ **removed 22.09**: the reviewer accepted the cut-paste bypass, so the second capability check it fed is gone and the wrapper with it. See the correction under finding 6.

Modified: `ensureCapability` wired as the **first statement** (before argument validation - a license error should beat a type error) in the ~20 methods from spec §05 task 2.2:
- NamedExpressions: `addNamedExpression`, `changeNamedExpression`, `removeNamedExpression`
- Clipboard: `copy`, `cut`, `paste` — Clipboard only. (The conditional **Crud** check described under finding 6 was removed 22.09.)
- Crud: `addRows`, `removeRows`, `addColumns`, `removeColumns`, `moveCells`, `moveRows`, `moveColumns`, `addSheet`, `removeSheet`, `clearSheet`, `setSheetContent`, `renameSheet`, `setCellContents`, `swapRowIndexes`, `setRowOrder`, `swapColumnIndexes`, `setColumnOrder`
- UndoRedo: `undo`, `redo`
- Batching: `batch`, `suspendEvaluation` — but **not** `resumeEvaluation`, see finding 4 below

`CustomFunctions` is not gated anywhere, per HF-307 decision D1. `ImportExport` has no methods to gate yet (HF-107).

**Where the line is drawn** (written into the `ensureCapability` JSDoc so a later change has to move it on purpose):
- **Gated** — mutations that create value: the sheet, the clipboard, the undo history, the named-expression set.
- **Not gated** — reads (`getCellValue`, `listNamedExpressions`, `getAllNamedExpressionsSerialized`, the `isItPossibleTo*` predicates) and teardown that only ever *removes* state (`clearClipboard`, `clearUndoStack`, `clearRedoStack`, `destroy`). Gating cleanup would let a restricted entitlement strand an integration mid-teardown while giving a licensee nothing. This mirrors gate B, which blocks *calling* a function rather than *reading* an already-computed value. There is a test pinning this, because "we chose not to gate cleanup" is otherwise indistinguishable from "we forgot these" — which is exactly how the reordering gap below survived the first pass, and how the cut/paste gap in finding 6 survived this one.

**Open question resolved:** the handoff left "is `getAllNamedExpressions` gated or read-only-exempt?" open. There's no such method today (it's `listNamedExpressions`, `getNamedExpression`, `getAllNamedExpressionsSerialized`) - I left all three ungated. This follows gate B's own precedent: it blocks *calling* a function, not *reading* a cell's already-computed value, so a restricted entitlement can still see named expressions that already exist. Flagging for reviewers in case the intended scope was broader.

Task 2.3: `BuildEngineFactory.ensureNamedExpressionsCapability` runs the same `allowsFeature(FeatureId.NamedExpressions)` check when the `namedExpressions` argument to `buildFromSheets`/`buildFromSheet`/`buildEmpty` (what `buildFromArray`/`buildFromSheets`/`buildEmpty` resolve to) is non-empty; an empty list is never checked. Deliberately **not** applied to `rebuildWithConfig` - that path re-serializes named expressions an already-built instance was already allowed to create (e.g. on `updateConfig()`), rather than accepting them fresh from a caller, so a later entitlement change must not retroactively break an existing instance's own state.

This ships without a real license-key payload adapter (PR 3), so every entitlement `Config` can produce today is unrestricted - `ensureCapability` and the build-time check are correct, independently-testable no-ops in production until that adapter lands, exactly like gate B in PR 1.

### Self-review findings (after the PR was first opened)

Recording these in the open rather than quietly amending, since several are things a reviewer should get to disagree with. Findings 4 and 5 came from Cursor Bugbot's inline comments and were confirmed here before fixing. Finding 6 came from an independent spec-to-ship review run against this PR specifically after it had already been open for review (a 5-dimension multi-agent pass: guard completeness, test-quality-by-mutation, the shipped build artifact, docs/CHANGELOG accuracy, and merge mechanics - each finding adversarially re-verified before being reported).

**1. The Crud gate was bypassable — fixed in `8a26e2be7`.** The first pass gated `moveRows`/`moveColumns` but missed four structurally identical methods: `swapRowIndexes`, `setRowOrder`, `swapColumnIndexes`, `setColumnOrder`. A restricted entitlement with no Crud grant could still permute every row and column in a sheet, which defeats the gate for a whole class of structural mutation. All four are now gated *in their own right* rather than relying on the `swap*` method that the `set*Order` pair delegates to — otherwise `set*Order`'s own `mappingFromOrder` validation would run first and a type error would beat the license error, violating task 2.2's first-statement rule.

**2. `ensureCapability` checks gate B only, never gate A — deliberate, and now an explicit invariant.** Today a missing or invalid key yields `#LIC!` in cells while the CRUD API keeps working; this PR preserves that exactly. The risk is forward-looking: if PR 3's key adapter resolves an invalid key to a *restricted* entitlement rather than an unrestricted one, every gated method here starts throwing and that becomes a silent breaking API change. This PR is what makes that reachable, so the invariant is documented at the guard itself. **Reviewers: worth confirming this is the intended split before PR 3 lands.**

**3. Known evidence gap in the task 2.3 tests.** The `BuildEngineFactory` gate is exercised by calling the private static directly, because the public path (`buildFromArray(sheet, config, namedExpressions)` → throws) *cannot* be reached today — the factories construct their own `Config`, and every `Config` currently resolves unrestricted. So Codecov's 100% patch coverage overstates the evidence for that one check: the integration path is unproven until PR 3 makes a restricted `Config` constructible. Flagging rather than papering over it.

**4. Gating `resumeEvaluation` could brick an engine — fixed in `c29a9ba40`.** Raised by Bugbot, confirmed. `resumeEvaluation` is the *only* exit from a suspended engine, and `_evaluationSuspended` survives `rebuildWithConfig`. So: suspend while Batching is granted → `updateConfig` produces an entitlement without Batching → the instance is permanently unusable, because every read throws `EvaluationSuspendedError` and the sole recovery path threw `LicenseCapabilityMissingError`. It is now ungated. `suspendEvaluation` and `batch` stay gated — those are the entry points that make the feature worth licensing, and if you cannot *enter* batching you can never extract value from it. Generalised into a rule on `ensureCapability` so it does not get re-added: **a capability check must never be reachable only on the way *out* of a state it let the caller into** — the same reasoning that leaves teardown ungated.

**5. One of PR 1's tests was vacuous — fixed in the paired tests PR.** Also raised by Bugbot. PR 1's custom-function-exemption test built the sheet with `'=CUSTOMFUNC()'` already in it, so the formula was evaluated during `buildFromArray` while the entitlement was still unrestricted; `restrictEngine` then changed nothing that `getCellValue` would re-read, and the assertion passed against a cached value. I verified this rather than assuming it — deleting the exemption from `Interpreter.ts` outright left the test **still passing**. It now writes the formula via `setCellContents` *after* restricting, matching the shape its sibling tests already had, and the same mutation now correctly fails it. The two new `resumeEvaluation` regression tests were verified the same way (re-adding the gate fails them).

**6. `paste` was a hard-gating bypass, plus a packaging gap and a documentation gap — found by an independent spec-to-ship review, all fixed.**
- **Bypass (HIGH).** `paste()` checked only `FeatureId.Clipboard`, but `CrudOperations.paste()` dispatches to `moveCells()` internally when the clipboard holds a cut - the exact cell-relocating mutation the public `moveCells()` requires **Crud** for. A `Clipboard`-only entitlement could reach it via `cut()`+`paste()`, with no Crud grant ever checked. Reproduced live (the move fully executed, no exception, under a Clipboard-only entitlement) before fixing. Fixed by adding `CrudOperations.isCutClipboard()` and additionally gating `paste()` on `Crud` when the clipboard holds a cut - still checked before argument validation. **Reversed 22.09 at the reviewer's request** ("Drop this check. We accept this bypass"): granting the clipboard is taken to grant what the clipboard does, so this route is deliberately not gated on Crud as well. The check, the wrapper and the stale `@throws` tag are gone; the paired spec now pins the acceptance - a clipboard-only entitlement relocates the cell - so the hole cannot be closed again by accident.
- **Packaging gap (HIGH).** `LicenseCapabilityMissingError` was never imported/exported from `src/index.ts`, so it was unreachable from the package's public entrypoint: `import {LicenseCapabilityMissingError} from 'hyperformula'` failed to compile (`TS2614`), the default-export static form was `undefined`, and no deep import worked either (the `exports` map in `package.json` only defines `.` and the i18n subpaths). A consumer had no supported way to catch this error by type - the existing test suite never caught it because it imports the class directly from `src/errors`, bypassing the public entrypoint entirely. Fixed by exporting it alongside the other ~30 error classes already there.
- **Documentation gap (MEDIUM).** None of the 27 gated methods carried a `@throws [[LicenseCapabilityMissingError]]` tag, breaking this file's own established per-method `@throws` convention (every *other* exception type already gets one). The class-level `@see` list on `LicenseCapabilityMissingError` itself also wrongly named `resumeEvaluation` (which finding 4 deliberately excludes) and omitted all 17 Crud methods. Fixed: all 27 gated instance methods plus the 3 static build factories now carry the tag, and the `@see` list is corrected and complete.

The same review also generalised finding coverage on the test side: mutation-testing every one of the 27 `ensureCapability` call sites found 17 with zero regression protection of their own (the "one test per feature group" strategy only pinned the group representative) - see the paired tests PR for the added coverage.

*Process note, since it generalises:* Bugbot's findings 4 and 5 were missed by the automated review tier that scrapes them (a regex looking for `Severity: Medium` against Bugbot's actual `**Medium Severity**`), which reported a clean PASS. Finding 6 was missed by both Bugbot and a prior manual review pass (which had only caught the narrower, single-method version of the test-coverage gap). Worth reading Bugbot's inline comments directly rather than trusting an aggregator, and worth an independent adversarial pass even after a PR looks clean.

### How did you test your changes?

- `tsc --noEmit` and `tsc -p tsconfig.test.json`: clean.
- `eslint` on all changed files: 0 errors (pre-existing warning categories only, none introduced by this change).
- Manual verification against the built engine (ts-node scratch script, not committed): every gated method throws `LicenseCapabilityMissingError` under a simulated restricted entitlement, an unrestricted entitlement (today's default) is unaffected, and `buildFromArray` with named expressions under the default unrestricted entitlement does not throw. Re-verified against a freshly rebuilt `commonjs` package for finding 6 specifically (the cut+paste fix and the `index.ts` export).
- Private test suite pushed: [hyperformula-tests#31](https://github.com/handsontable/hyperformula-tests/pull/31) - `unit/license/public-api-guards.spec.ts` (42 tests: one positive + one negative per feature group, a validation-order test, the build-time bypass pair, five covering the reordering gap, one pinning the ungated-cleanup rule, two for the `resumeEvaluation` deadlock, one dedicated throw test per previously-group-only-covered method (15 of them), the cut/paste bypass control+regression pairs, and one proving `LicenseCapabilityMissingError` is reachable from the public entrypoint) plus fixes to PR 1's `unit/licence.spec.ts` (the test-helper interaction described in that PR, and the vacuous exemption test in finding 5).
- **Mutation-checked, not just green:** the two `resumeEvaluation` tests, the repaired custom-function test, the cut+paste bypass fix itself, and a sample of the newly-added per-method tests (`removeRows`, `changeNamedExpression`) were each verified by reverting the corresponding source line and confirming the new test fails, then restoring.
- Full suite against the paired branch: **511 suites / 6373 tests, 6370 passed, 3 pre-existing skips, 0 failures.**

### Types of changes
- [ ] Breaking change
- [x] New feature or improvement
- [x] Bug fix
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
1. HF-307 (internal tracker; no public GitHub issue for this one)

### Checklist:
- [x] I have reviewed the guidelines about Contributing to HyperFormula and I confirm that my code follows the code style of this project.
- [ ] I have signed the Contributor License Agreement. *(please confirm/attach on your end - I can't verify this from here)*
- [ ] My change is compliant with the OpenDocument standard. *(N/A - no worksheet function behaviour changes in this PR)*
- [ ] My change is compatible with Microsoft Excel. *(N/A - same reason)*
- [ ] My change is compatible with Google Sheets. *(N/A - same reason)*
- [ ] I described my changes in the CHANGELOG.md file. *(intentionally not done - internal-only change, no user-visible behaviour yet; a CHANGELOG entry lands with the PR that actually activates the gates for customers)*
- [x] My changes require a documentation update. *(done - @throws tags and the @see list on LicenseCapabilityMissingError, see finding 6)*
- [ ] My changes require a migration guide. *(no)*

---

> [!NOTE]
> **Medium Risk**
> Wide API surface change: many mutation entry points can throw when restricted entitlements land in PR 3; deliberate exclusions (ungated `resumeEvaluation`, Clipboard-only paste) affect runtime behavior under partial licenses.
> 
> **Overview**
> Adds **license entitlement guards (HF-307 PR 2)** on the public HyperFormula API so restricted entitlements throw `LicenseCapabilityMissingError` instead of mutating the engine.
> 
> A private `ensureCapability` runs as the **first statement** on gated instance methods when `isLicenseGateActive` is true: **Crud** (sheet/cell/row/column moves and structural edits), **UndoRedo**, **Clipboard** (`copy`/`cut`/`paste`—paste is Clipboard-only even when a cut relocates cells), **Batching** (`batch`, `suspendEvaluation` only; **`resumeEvaluation` stays ungated** so entitlement changes cannot strand a suspended instance), and **NamedExpressions** (add/change/remove). Reads, `isItPossibleTo*` helpers, and teardown (`clearClipboard`, undo/redo stack clears, `destroy`) stay ungated.
> 
> **Build-time:** `BuildEngineFactory.ensureNamedExpressionsCapability` rejects non-empty initial `namedExpressions` without the feature on `buildFromSheets`/`buildFromSheet`/`buildEmpty`, but not on `rebuildWithConfig`. The new error is exported from `src/index.ts`, with `@throws` docs on affected methods. Until PR 3 supplies restricted entitlements, these checks are no-ops for today’s default configs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 690ef69d758a09c51e7e5ab97b203e137c52e632. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [HF-107 Import/export files (XLSX, CSV)](https://app.clickup.com/t/9015210959/HF-107)
Task linked: [HF-307 Implement feature packages and add-ons in HF](https://app.clickup.com/t/9015210959/HF-307)

File: src/HyperFormula.ts [hunk 1]
```diff
@@ -3795,6 +3818,7 @@ export class HyperFormula implements TypedEmitter {
    * @category Batch
    */
   public resumeEvaluation(): ExportedChange[] {
+    this.ensureCapability(FeatureId.Batching)
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/HyperFormula.ts]: Confirmed and fixed in c29a9ba40 — good catch, this was a real deadlock.

Verified the mechanism: `_evaluationSuspended` is never reset by `rebuildWithConfig` (it only reassigns `_config`, `_dependencyGraph`, `_crudOperations`, …), so an instance suspended while `Batching` was granted and then reconfigured without it had no escape at all — every read throws `EvaluationSuspendedError` and the only exit threw `LicenseCapabilityMissingError`.

`resumeEvaluation` is now ungated. `suspendEvaluation` and `batch` stay gated: those are the entry points that make batching worth licensing, and a caller who cannot *enter* batching can never extract value from it, so nothing is lost by letting them leave. Generalised into a rule on `ensureCapability` so it does not get re-added — a capability check must never be reachable only on the way *out* of a state it let the caller into, which is the same reasoning that already left `clearClipboard`/`clearUndoStack`/`clearRedoStack` ungated.

Two regression tests added in handsontable/hyperformula-tests#31, both mutation-checked: re-adding the gate fails them.

File: src/HyperFormula.ts [hunk 2]
```diff
@@ -4767,6 +4848,45 @@ export class HyperFormula implements TypedEmitter {
...
+   * Note this checks gate B (entitlement) only, never gate A (key validity). That asymmetry
+   * with the interpreter's gate B - which checks key validity first - is deliberate: it keeps
+   * today's behaviour for a missing or invalid key, where formulas yield `#LIC!` but the CRUD
+   * API keeps working. A later PR that resolves an invalid key to a *restricted* entitlement
+   * rather than an unrestricted one would silently turn that into a breaking API change.
+   *
+   * @internal
+   */
+  private ensureCapability(feature: FeatureId): void {
+    if (!this._config.isLicenseGateActive) {
```
Comment (Tobiadefami) [inline src/HyperFormula.ts:4882]: The active-gate entitlement checks are covered, but I couldn’t find a test for this short-circuit. A case with `isLicenseGateActive === false` should confirm that the guarded operation continues without consulting its capabilities.

File: src/BuildEngineFactory.ts [hunk 3]
```diff
@@ -50,23 +52,44 @@ export type EngineState = {
 export class BuildEngineFactory {
   public static buildFromSheets(sheets: Sheets, configInput: Partial<ConfigParams> = {}, namedExpressions: SerializedNamedExpression[] = []): EngineState {
     const config = new Config(configInput)
+    this.ensureNamedExpressionsCapability(config, namedExpressions)
```
Comment (Tobiadefami) [inline src/BuildEngineFactory.ts:55]: The entitlement tests currently call `ensureNamedExpressionsCapability` directly via `as any`. That covers the helper’s logic but not these three call sites. Exercising `HyperFormula.buildFromArray`, HyperFormula.buildFromSheets`, and `HyperFormula.buildEmpty` with initial named expressions under a restricted entitlement would cover the public build paths.

**Review by Tobiadefami — COMMENTED** [review]
Review focused on entitlement guard behavior and end-to-end coverage.

File: src/HyperFormula.ts (hunk 2 above)
Comment (marcin-kordas-hoc) [reply src/HyperFormula.ts:4882]: You were right that the branch had no cover at all — every other test in that file forces the gate ON. Tests commit 46e357f (hyperformula-tests, same branch) adds *gates nothing while the license gate is inactive, whatever the capabilities say*: gate inactive, capabilities granting nothing, and five guarded public methods must all still pass.

File: src/BuildEngineFactory.ts (hunk 3 above)
Comment (marcin-kordas-hoc) [reply src/BuildEngineFactory.ts:55]: Good catch, and worth spelling out why the fix is not in this PR. At PR 2 there is no typed-key reader yet, so no public call can produce a restricted entitlement — the `as any` handle was the only one that existed, which is exactly the gap you spotted: deleting all three call sites left the suite green. The end-to-end version lives on PR 3, in `unit/license/typed-key-resolution.spec.ts` — `buildFromArray`, `buildFromSheets` and `buildEmpty` each built against a key granting only `feat:clipboard`, plus two counter-cases (key grants the feature; no named expression passed).

File: src/HyperFormula.ts [hunk 4]
```diff
@@ -2443,6 +2483,16 @@ export class HyperFormula implements TypedEmitter {
...
+    this.ensureCapability(FeatureId.Clipboard)
+    // Pasting a CUT moves cells across the sheet - the same mutation the public moveCells()
+    // requires Crud for - so a Clipboard-only entitlement must not reach it this way. Pasting a
+    // COPY only duplicates values/formulas and stays Clipboard-only, correctly. Checked before
+    // argument validation, same as every other ensureCapability call (HF-307 spec-to-ship review,
+    // 18.08: found as a hard-gating bypass - a Clipboard-only entitlement could cut() then
+    // paste() to relocate cells without Crud ever being granted).
+    if (this._crudOperations.isCutClipboard()) {
+      this.ensureCapability(FeatureId.Crud)
+    }
```
Comment (sequba) [inline src/HyperFormula.ts]: Drop this check. We accept this bypass

**Review by sequba — CHANGES_REQUESTED** [review]

File: src/HyperFormula.ts (hunk 4 above)
Comment (marcin-kordas-hoc) [reply src/HyperFormula.ts]: @sequba Done. Dropped, along with the `CrudOperations.isCutClipboard()` wrapper that existed only to reach it; the clipboard-state helper it delegated to predates this work and stays.

One thing I did not just delete: the paired suite used to pin the refusal, and it now pins the acceptance instead — pasting a cut relocates cells on a clipboard-only entitlement, and the test asserts the cell really moved. Without that, the next person to run a gating audit finds the same hole and "fixes" it again, because nothing in the code distinguishes a decision from an oversight.

A review harness then caught what I had missed: the `@throws` tag on the same method still promised the Crud refusal, three lines above the comment explaining why it is not there. Corrected in the same round.
